File: Source/typescript/features/for_FeaturesParser/when_parsing_with_two_features_with_one_toggle_each.ts
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import { FeaturesParser } from '../FeaturesParser';

const first_feature = 'my.first.feature';
const second_feature = 'my.second.feature';
const first_feature_description = 'My first feature';
const second_feature_description = 'My second feature';



const json: string =
    `{` +
    `   "${first_feature}": {` +
    `       "description": "${first_feature_description}",` +
    `       "toggles": [{` +
    `           "type": "Boolean",` +
    `           "isOn": false` +
    `       }]` +
    `   },` +
    `   "${second_feature}": {` +
    `       "description": "${second_feature_description}",` +
    `       "toggles": [{` +
    `           "type": "Boolean",` +
    `           "isOn": true` +
    `       }]` +
    `   }` +
    `}`;



describe('when parsing with two features with one toggle each', () => {

    const parser = new FeaturesParser();

    const result = parser.parse(json);

    it('should contain first feature', () => result.has(first_feature).should.be.true);
    it('should hold first features description', () => result.get(first_feature)?.description.should.equal(first_feature_description));
    it('should have one toggle for first feature', () => result.get(first_feature)?.toggles.length.should.equal(1));
    it('should have toggle for first feature set to false', () => result.get(first_feature)?.toggles[0].isOn.should.be.false);
    it('should contain second feature', () => result.has(second_feature).should.be.true);
    it('should hold second features description', () => result.get(second_feature)?.description.should.equal(second_feature_description));
    it('should have one toggle for second feature', () => result.get(second_feature)?.toggles.length.should.equal(1));
    it('should have toggle for second feature set to true', () => result.get(second_feature)?.toggles[0].isOn.should.be.true);
});

File: Specifications/DotNET/Backend/Features/for_FeaturesParser/when_parsing_with_two_features_with_one_toggle_each.cs
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Linq;
using Machine.Specifications;

namespace Dolittle.Vanir.Backend.Features.for_FeaturesParser
{
    public class when_parsing_with_two_features_with_one_toggle_each
    {
        const string first_feature = "my.first.feature";
        const string second_feature = "my.second.feature";

        const string json =
        "{" +
        "   \"my.first.feature\": {" +
        "       \"description\": \"The first feature\"," +
        "       \"toggles\": [{" +
        "           \"type\": \"Boolean\"," +
        "           \"isOn\": false" +
        "       }]" +
        "   }," +
        "   \"my.second.feature\": {" +
        "       \"description\": \"The second feature\"," +
        "       \"toggles\": [{" +
        "           \"type\": \"Boolean\"," +
        "           \"isOn\": true" +
        "       }]" +
        "   }" +
        "}";

        static FeaturesParser parser;
        static Features result;

        Establish context = () => parser = new();

        Because of = () => result = parser.Parse(json);

        It should_contain_first_feature = () => result.ContainsKey(first_feature).ShouldBeTrue();
        It should_hold_first_features_description = () => result[first_feature].Description.ShouldEqual("The first feature");
        It should_have_one_toggle_for_first_feature = () => result[first_feature].Toggles.Count().ShouldEqual(1);
        It should_have_toggle_for_first_feature_set_to_false = () => result[first_feature].Toggles.First().IsOn.ShouldBeFalse();
        It should_contain_second_feature = () => result.ContainsKey(second_feature).ShouldBeTrue();
        It should_hold_second_features_description = () => result[second_feature].Description.ShouldEqual("The second feature");
        It should_have_one_toggle_for_second_feature = () => result[second_feature].Toggles.Count().ShouldEqual(1);
        It should_have_toggle_for_second_feature_set_to_true = () => result[second_feature].Toggles.First().IsOn.ShouldBeTrue();
    }
}
